Connect: Adjust UI after making useDocumentTerminal resolve hostname

ravicious opened the pull request:
Part of #21964.

This PR takes care of a couple of UI changes after the `useDocumentTerminal` changes introduced in the previous PR, #22181.

The biggest of them is the refactoring of error handling of `useDocumentTerminal` to provide more user-friendly flow of retrying PTY creation when something goes wrong. This had to be done because now the PTY creation might fail not only due to problems with actually setting up the PTY but it might also fail when the request to resolve the hostname fails.

Best reviewed commit-by-commit, I left commit messages where appropriate.

ravicious committed:
Do not create connections for tsh ssh docs with no serverUri

ravicious committed:
Minor type adjustments after DocumentTshNode changes

ravicious committed:
Show file transfer bar while hostname is being resolved

ravicious committed:
Set doc status to connecting when creating doc from connection

This should be changed a long time ago. Currently, there's no other place
which depends on the status being disconnected. That's going to change in
upcoming commits as I'm going to reuse it.

In the meantime, let's change it to 'connecting' so that the progress bar
is correctly shown when creating a new tsh ssh doc based on a connection
from the list.

ravicious committed:
useDocumentTerminal: Improve error handling & reconnection flow

Before this change, useDocumentTerminal had a rather convoluted error
handling where different the actual error handling was split across
different functions and errors would be treated differently depending on
where they happened.

This commit makes it so that any error thrown by startTerminalSession
will make the useAsync attempt fail. A failed attempt will make
DocumentTerminal display a DocumentReconnect component which allows
the user to either retry resolving the hostname or just simply attempt
to create a PTY again.

Since the `disconnected` value of the `status` field was not used, it got
replaced with `error`.

ravicious commented on web/packages/teleterm/src/ui/services/connectionTracker/connectionTrackerService.test.ts:
```diff
@@ -97,14 +87,119 @@ it('removeItemsBelongingToRootCluster removes connections', () => {
   ]);
 });
 
 it('updates the port of a gateway connection when the underlying doc gets updated', () => {
```
git shows a weird diff here but basically what I've done is I extracted the common setup out of this test case into `getTestSetupWithMockedDocuments` and added two new test cases.

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -31,7 +31,11 @@ import useDocTerminal, { Props } from './useDocumentTerminal';
+    doc.kind === 'doc.terminal_tsh_node' &&
+    'serverUri' in doc &&
+    doc.status === 'disconnected'
+  ) {
```
`DocumentReconnect` is going to get completely rewritten in another commit in this PR. This change is just to satisfy the type system.

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -62,59 +117,15 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
+      {$fileTransfer}
       {ptyProcess && (
-        <>
-          {doc.kind === 'doc.terminal_tsh_node' && (
```
Previously, the file transfer UI would be shown only after the PTY got created. This doesn't work for the doc type as the PTY is created only after `loginHost` is resolved.

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -124,9 +132,60 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
+      };
+    }
+    default:
+      assertUnreachable(doc);
```
As I said before, I keep going back and forth on throwing within UI code.

The ultimate problem is that if we add a new doc type and then the user downgrades the app while having the new doc in the state, this code will crash the app. It'd be completely okay to keep it _if_ we had zod throw out any unsupported docs from app_state.json before loading them into the actual app state (which we currently don't do).

In the end, I think I prefer type safety in code vs being careful when adding new docs vs not having type safety. As in, with the current code the type system will return an error if someone extends `DocumentTerminal` without addressing these exhaustive checks.

gzdunek replied:
Are you sure it would crash the app? Isn't it that an unsupported document is caught by `DocumentRenderer`?
I mean:
```tsx
default:
        return (
          <Document visible={visible}>
            Document kind "{doc.kind}" is not supported
          </Document>
        );
```

ravicious replied:
> Are you sure it would crash the app? Isn't it that an unsupported document is caught by `DocumentRenderer`?

Ah, you're right!

gzdunek reviewed (commented):
I tested it, and the UX is much nicer comparing to the previous approach :)

ravicious committed:
DocumentTshNodeWithServerId: Make loginHost undefined

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -54,6 +55,60 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
+      $fileTransfer = (
+        <FileTransferContextProvider>
+          <FileTransferActionBar isConnected={doc.status === 'connected'} />
+          <FileTransfer
```
> I'm pretty sure we could achieve the same effect by wrapping only the `FileTransfer` component:

Good call!

ravicious replied:
I kept the `$fileTransfer` var so that it's easier to scan the structure of the final component.

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -124,9 +132,60 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
   );
 }
+
+function DocumentReconnect(props: {
```
> IMO this component is big enough to have its own file. I don't see any benefits from keeping it here.

I don't think it's particularly big, it's also not reused anywhere else and I like keeping [things that change together](https://programmingisterrible.com/post/176657481103/repeat-yourself-do-more-than-one-thing-and), well, together physically as well. The only reason for which `DocumentReconnect` could ever change is if a change in `DocumentTerminal` had to be propagated down to `DocumentReconnect`. This way you have one less file to hunt down.

Also gonna do a small appeal to authority here. 😏

* https://twitter.com/dan_abramov/status/1439304400464449536
* https://twitter.com/dan_abramov/status/1220029098543788032
* https://twitter.com/dan_abramov/status/1230841812115501056

gzdunek replied:
Ok!

ravicious committed:
Simplify file transfer in DocumentTerminal

gzdunek commented on web/packages/teleterm/src/ui/DocumentTerminal/index.ts:
```diff
@@ -14,5 +14,5 @@ See the License for the specific language governing permissions and
 
-import DocumentTerminal from './DocumentTerminal';
-export default DocumentTerminal;
+import { DocumentTerminal } from './DocumentTerminal';
```
Nit:
```ts
export { DocumentTerminal } from './DocumentTerminal';
```

gzdunek commented on web/packages/teleterm/src/ui/services/connectionTracker/connectionTrackerService.test.ts:
```diff
@@ -97,14 +87,119 @@ it('removeItemsBelongingToRootCluster removes connections', () => {
+  expect((connection as TrackedGatewayConnection).port).toBe('54321');
+});
+
+it('creates a connection foc doc.terminal_tsh_node docs with serverUri', () => {
```
Typo in "foc"

gzdunek reviewed (commented):
Looks good, but I would like to come back to it on Monday

gzdunek commented on web/packages/teleterm/src/ui/DocumentTerminal/useReconnect.ts:
```diff
@@ -22,7 +22,7 @@ import { useAppContext } from 'teleterm/ui/appContextProvider';
 import * as types from 'teleterm/ui/services/workspacesService';
```
`useReconnect` can be removed.

ravicious replied:
Good catch, thanks.

ravicious committed:
Explain existence of loginHost on DocumentTshNodeWithServerId

ravicious committed:
Add type guards for DocumentTshNode subtypes

ravicious committed:
Simplify DocumentTerminal reexport

ravicious committed:
Fix typo in connectionTrackerService.test.ts

ravicious committed:
Remove useReconnect from DocumentTerminal

ravicious committed:
useDocumentTerminal: Move startTerminal error handling inside useAsync

Both useEffect and reconnect were calling startTerminal but reconnect
wasn't checking the error of startTerminal. This meant that if there was
an error on reconnect, the status of the doc was never updated.

ravicious commented:
Merged into #21964 in 5d5a069c07d.
